### PR TITLE
email functionality

### PR DESCRIPTION
**Pull request by shindym**
- Email functionality with database
- generate pdf report in class

### CODE REVIEW

File: weekly_report/generate_report.py [hunk 1]
```diff
@@ -224,168 +228,227 @@ def get_website_names(self) -> list[str]:
...
-    for developer in website_stats_retriever.top_three_developers():
-        canvas_obj.drawString(50, x, f'- {developer}')
-        x -= 20
-    canvas_obj.setFillColorRGB(0.6, 0.70, 30.0)
-    canvas_obj.drawString(300, 20, f'{canvas_obj.getPageNumber()}')
+class ReportMaker():
+    def __init__(self, config) -> None:
+        self.config = config
+
+    def generate_summary_text(self, canvas_obj, sum):
```
Comment (annalisev) [inline weekly_report/generate_report.py:235]: doc string here

File: weekly_report/generate_report.py [hunk 2]
```diff
@@ -224,168 +228,227 @@ def get_website_names(self) -> list[str]:
...
-    canvas_obj.drawString(50, 200, 'Top 3 Developers:')
-    x = 180
-    canvas_obj.setFillColorRGB(0.6, 0.70, 35.0)
-    for developer in website_stats_retriever.top_three_developers():
-        canvas_obj.drawString(50, x, f'- {developer}')
-        x -= 20
-    canvas_obj.setFillColorRGB(0.6, 0.70, 30.0)
-    canvas_obj.drawString(300, 20, f'{canvas_obj.getPageNumber()}')
+class ReportMaker():
+    def __init__(self, config) -> None:
```
Comment (annalisev) [inline weekly_report/generate_report.py:232]: doc string here?

File: weekly_report/generate_report.py [hunk 3]
```diff
@@ -224,168 +228,227 @@ def get_website_names(self) -> list[str]:
...
-    canvas_obj.setFillColorRGB(0.6, 0.50, 30.0)
-    canvas_obj.drawString(50, 200, 'Top 3 Developers:')
-    x = 180
-    canvas_obj.setFillColorRGB(0.6, 0.70, 35.0)
-    for developer in website_stats_retriever.top_three_developers():
-        canvas_obj.drawString(50, x, f'- {developer}')
-        x -= 20
-    canvas_obj.setFillColorRGB(0.6, 0.70, 30.0)
-    canvas_obj.drawString(300, 20, f'{canvas_obj.getPageNumber()}')
+class ReportMaker():
```
Comment (hhm970) [inline weekly_report/generate_report.py:231]: Good use of OOP to manage and organise your functions!

File: weekly_report/generate_report.py [hunk 4]
```diff
@@ -224,168 +228,227 @@ def get_website_names(self) -> list[str]:
...
+        canvas_obj.drawString(50, 200, 'Top 3 Developers:')
+        x = 180
+        canvas_obj.setFillColorRGB(0.6, 0.70, 35.0)
+        for developer in website_stats_retriever.top_three_developers():
+            canvas_obj.drawString(50, x, f'- {developer}')
+            x -= 20
+        canvas_obj.setFillColorRGB(0.6, 0.70, 30.0)
+        canvas_obj.drawString(300, 20, f'{canvas_obj.getPageNumber()}')
+
+    def generate_report(self):
```
Comment (annalisev) [inline weekly_report/generate_report.py:371]: docstring

Comment (Eibironke) [inline weekly_report/generate_report.py:371]: Docstring on functions if possible

File: weekly_report/generate_report.py [hunk 5]
```diff
@@ -224,168 +228,227 @@ def get_website_names(self) -> list[str]:
...
+        return connect(
+            dbname=config["DB_NAME"],
+            user=config["DB_USER"],
+            password=config["DB_PASSWORD"],
+            host=config["DB_HOST"],
+            port=config["DB_PORT"],
+            cursor_factory=RealDictCursor
+        )
+
+    def get_subscription_emails(self) -> list[str]:
```
Comment (hhm970) [inline weekly_report/generate_report.py:412]: Make sure to include a `connection` object as an input

File: weekly_report/generate_report.py [hunk 6]
```diff
@@ -224,168 +228,227 @@ def get_website_names(self) -> list[str]:
...
+        c.showPage()
+        # GOG
+        self.individual_summary(c, gog, 2, "GOG")
+        c.showPage()
+        # EPIC
+        self.individual_summary(c, epic, 3, "Epic")
+        c.save()
+
+
+class Alerter():
```
Comment (Eibironke) [inline weekly_report/generate_report.py:397]: Classes also need docstrings

File: weekly_report/generate_report.py [hunk 7]
```diff
@@ -224,168 +228,227 @@ def get_website_names(self) -> list[str]:
...
+            RawMessage={
+                'Data': message.as_string()
+            }
+        )
 
 
 if __name__ == "__main__":
     load_dotenv()
-    generate_pdf(os.environ)
+    r = ReportMaker(os.environ)
```
Comment (Eibironke) [inline weekly_report/generate_report.py:450]: better variable names maybe?

**Review by Eibironke — CHANGES_REQUESTED** [review]

**Review by hhm970 — COMMENTED** [review]
Make sure to ensure docstrings and type hints in all functions. Otherwise the code looks efficiently programmed and organised - well done!

File: weekly_report/generate_report.py (hunk 3 above)
Comment (Eibironke) [reply weekly_report/generate_report.py:231]: Yh really good

**Review by annalisev — APPROVED** [review]
Looks really good, thank you for adding docstrings!

**Review by Eibironke — APPROVED** [review]
looks good

**Pull request merged**